[CSHARP-969] - OpenTelemetry tracing RFC (#1)
simaoribeiro/csharp-driver: doc/features/open-telemetry/tracing-rfc.md

> - Feature Name: Add OpenTelemetry Traces

Author: TODO: Include team in the commit when squashing and rebasing

Author: Co Authored and reviwed-by included in git message

For the .NET ecosystem, there are available implementations regarding the export of telemetry data in client-side calls in some major database management systems, being them community lead efforts ([SqlClient](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.SqlClient), [MongoDB](https://github.com/jbogard/MongoDB.Driver.Core.Extensions.OpenTelemetry)), or native implementations ([Elasticsearch](https://github.com/elastic/elasticsearch-net/blob/main/src/Elastic.Clients.Elasticsearch/Client/ElasticsearchClient.cs#L183)).

Cassandra has a community developed package available in the [opentelemetry-dotnet-contrib](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Cassandra) that exports metrics but not traces. This proposal to include traces in the native Cassandra C# driver will allow the developers to have access to database operations when analyzing the requests that are handled in their systems when it includes Cassandra calls.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## [Traces](https://opentelemetry.io/docs/concepts/signals/traces/)

> As mentioned in [*Motivation*](#motivation), traces allows the developers to understand the full "path" a request takes in the application and navigate through a microservice architecture. Traces include [Spans](https://opentelemetry.io/docs/concepts/signals/traces/#spans) which are unit of works or operation in the ecosystem that include the following information:

Reviewer: 💅 related with this statement "and navigate through a microservice architecture". Tracing is still valuable in non microservices architecture. So maybe don't bind tracing to that architecture only.

Author: Done, please review

Reviewer: looks good.

### Span attributes

This implementation includes, by default, the **required** attributes for Database, and Cassandra spans.\
`server.address` and `server.port`, despite only **recommended**, are included to give information regarding the client connection.\
`db.statement` is optional given that this attribute may contain sensitive information:

| Attribute  | Description  | Type | Required | Supported Values|
|---|---|---|---|---|
| span.kind | Describes the relationship between the Span, its parents, and its children in a Trace. | string | true | client |
| db.system | An identifier for the database management system (DBMS) product being used. | string | true | cassandra |
| db.name | The keyspace name in Cassandra. | string | true | *keyspace in use* |
| db.statement | The database statement being executed. | string | false | *database statement in use* |
| server.address | Name of the database host. | string | true | e.g.: example.com; 10.1.2.80; /tmp/my.sock |
| server.port | Server port number. Used in case the port being used is not the default | int | false | e.g.: 9445 |

## Usage

### Package installation

> The OpenTelemetry implementation is included in the package `CassandraCSharpDriver.OpenTelemetry`.

Reviewer: Since we are referring to a proposal for future implementation, we should probably rephrase it differently:

```suggestion
The OpenTelemetry implementation will be included in the package `CassandraCSharpDriver.OpenTelemetry`.
```
Does it make sense?

Author: Done, please review

Reviewer: 👍

### Exporting Cassandra activity

> The extension method `AddOpenTelemetryInstrumentation()` is available in the cluster builder, so the activity can be exported for database operations:

Reviewer: Also related to this [comment](https://github.com/simaoribeiro/csharp-driver/pull/1/files#r1441584971) but using the present tense creates some confusion if this is available right now or if it is referring to the proposed change.

Author: Done, please review

Reviewer: 👍

```csharp
var cluster = Cluster.Builder()
    .AddContactPoint("127.0.0.1")
    .WithSessionName("session-name")

>     .AddOpenTelemetryInstrumentation()

Reviewer: The open-telemetry-contrib for [cassandra](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Cassandra) already defines a `.WithOpenTelemetryMetrics()`.  By naming this `.AddOpenTelemetryInstrumentation()` couldn't this mistakenly imply that it will enable all the instrumentation for Cassandra (metrics, tracing)?

Author: Good question. Metrics are not being exported or created by any standard and, in [future possibilities](https://github.com/simaoribeiro/csharp-driver/blob/793dfec7b6262208094b14c133b42871060e3780/doc/features/open-telemetry/tracing-rfc.md#future-possibilities) I mentioned that metrics could be exported from OTEL in the future.
However, it makes sense that the extension method should have a more explicit name. Maybe `AddOpenTelemetryTraceInstrumentation()`? What do you say?

Author: Done, please review

Reviewer: I agree with `AddOpenTelemetryTraceInstrumentation` in order to prevent miss judgement around what that is doing. Unless the implementation of other instrumentation features are planned for the "near" future.

Nevertheless, their input on this is going to be more valuable to understand the best naming.

### Tracer observers

The tracing implementation will use the already defined factory interfaces for the creation of observer factories, so the following classes will be created:

```csharp
internal class TracerObserverFactoryBuilder : IObserverFactoryBuilder
```

```csharp
internal class TracerObserverFactory : IObserverFactory
```

When instantiating the factories that will culminate in the creation if the `TracesRequestObserver`, the implementation of `IDriverTrace` will be passed as parameter. This interface will define the following methods:

```csharp
public interface IDriverTrace
{
    void OnStart(IStatement statement);

    void OnError(Exception ex);
}
```

and will be used in the `TracesRequestObserver`:

```csharp
internal class TracesRequestObserver : IRequestObserver

(...)

public void OnRequestError(Host host, RequestErrorType errorType, RetryDecision.RetryDecisionType decision)
{
}

public void OnRequestFinish(Exception exception)
{
    _driverTrace.OnError(exception);
}

public void OnRequestStart(IStatement statement)
{
    _driverTrace.OnStart(statement);
}

public void OnSpeculativeExecution(Host host, long delay)
{
}
```

### Public API

> The `Builder` will include a new method named `WithTracer` that will include an `IDriverTrace` instance being passed as parameter. This method can be used by anyone to pass their tracer implementation and will be used by the extension method `AddOpenTelemetryInstrumentation()` mentioned in the previous section.

Reviewer: What is the use case for someone to provide their own `IDriverTrace` implementation?

Author: I would say, if someone wants to implement their own instrumentation like not using OTEL or prefer to create their own activity with their own attributes, it would make sense.
Also, using IDriverTracer is a way to separate the interface from its implementation without the use of a provider builder, just like Cassandra.AppMetrics implementation has

Reviewer: Ok 👍

- Feature Name: Add OpenTelemetry Traces
- Start Date: 2023-12-27

> 

Reviewer 2: One topic that seems not be covered in this RFC is the [IStatement.IsTracing](https://github.com/simaoribeiro/csharp-driver/blob/bc6863efd17a01c26c46369248205c1a35a5a6e0/src/Cassandra/IStatement.cs#L53) property. 

I am highlighting this because I remember that this property could lead to certain doubts with the OpenTelemetry implementation.

Is it not a concern anymore?

To avoid parsing the statement, the **span name** in this implementation has the **keyspace name** as its value.

### Span attributes

This implementation will include, by default, the **required** attributes for Database, and Cassandra spans.\
`server.address` and `server.port`, despite only **recommended**, are included to give information regarding the client connection.\
`db.statement` is optional given that this attribute may contain sensitive information:

| Attribute  | Description  | Type | Required | Supported Values|
|---|---|---|---|---|
| span.kind | Describes the relationship between the Span, its parents, and its children in a Trace. | string | true | client |
| db.system | An identifier for the database management system (DBMS) product being used. | string | true | cassandra |
| db.name | The keyspace name in Cassandra. | string | true | *keyspace in use* |
| db.statement | The database statement being executed. | string | false | *database statement in use* |
| server.address | Name of the database host. | string | true | e.g.: example.com; 10.1.2.80; /tmp/my.sock |

> | server.port | Server port number. Used in case the port being used is not the default | int | false | e.g.: 9445 |

Reviewer 3: ```suggestion
| server.port | Server port number. Used in case the port being used is not the default. | int | false | e.g.: 9445 |
```

```csharp
public Builder WithTracer(IDriverTrace trace)
{
    _driverTracer = trace;
    return this;
}
```

### Changes to Cassandra core

#### IRequestObserver

As the tracer implementation needs information included in `IStatement`, it will be necessary to change the definition of the `IRequestObserver.OnRequestStart()` to include it as a parameter.

```csharp
internal interface IRequestObserver
{
    (...)

    void OnRequestStart(IStatement statement);

    (...)
}
```

#### Composite observers and factories
[composite-observers-and-factories]: #composite-observers-and-factories

The core package includes observers that are instantiated through the configuration points in the driver, being the *Configuration* class one example:

```csharp
internal Configuration(...)
{
    (...)
    
    ObserverFactoryBuilder = observerFactoryBuilder ?? (MetricsEnabled ? (IObserverFactoryBuilder)new MetricsObserverFactoryBuilder() : new NullObserverFactoryBuilder());

    (...)
}
```

> As it is necessary to have multiple observers being triggered on driver'actions, this proposal includes a composite pattern that will aggregate multiple observer instances for that work.

Reviewer 4: ```suggestion
As it is necessary to have multiple observers being triggered on driver's actions, this proposal includes a composite pattern that will aggregate multiple observer instances for that work.
```

```csharp
internal Configuration(...)
{
    (...)
     ObserverFactoryBuilder = new CompositeObserverFactoryBuilder(
                new MetricsObserverFactoryBuilder(MetricsEnabled),
                new TracerObserverFactoryBuilder(driverTracer));
    (...)
}
```

As it is possible to be seen, the validation to construct the `MetricsRequestObserver` will be moved to the `MetricsObserverFactoryBuilder` instead of being done in the configuration, e.g.:

```csharp
public IObserverFactory Build(IMetricsManager manager)
{
    return this.isEnabled ? new MetricsObserverFactory(manager) : NullObserverFactory.Instance;
}

```

#### Cassandra.Requests

`TcsMetricsRequestResultHandler` now receives `IStatement` as a constructor parameter, so it can be passed to the updated `IRequestObserver.OnRequestStart(IStatement statement)`.

As so, the instantiation of the `IRequestResultHandler` in `RequestHandler` class will also change to include the statement as parameter_

```csharp
_requestResultHandler = new TcsMetricsRequestResultHandler(_requestObserver, statement);
```

# Drawbacks
[drawbacks]: #drawbacks

> Although the implementation suggested in this document tries to minimize changing the core package, the inclusion of new observers being called will increase the complexity and the performance of the driver.

Reviewer 4: Will it increase the performance or it will decrease it?

The method [`RecordException()`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Trace/ActivityExtensions.cs#L81) is simple enough to be replicated and can be included in the `Cassandra.OpenTelemetry` project. As a downside, that means the Cassandra project will need to keep up with the best practices on how to report exceptions for tracing, and will include code duplication. As an upside, it is possible to target .NET Framework `4.6.1`, and the reference to `OpenTelemetry.Api` is not needed which also allows the Cassandra project to not have a possible deprecated dependency depending on the evolution of reference package.

## Using OpenTelemetry.SemanticConventions package

The [semantic conventions](https://opentelemetry.io/docs/specs/semconv/) are a fast evolving reference that "define a common set of (semantic) attributes which provide meaning to data when collecting, producing and consuming it.".\

> As it changes can be hard to follow, the .NET project includes a package named [`OpenTelemetry.SemanticConventions`](https://www.nuget.org/packages/OpenTelemetry.SemanticConventions/1.0.0-rc9.9) that maps the attributes defined in the conventions to a .NET project. Using this package will allow the Cassandra project to have its tracing attributes up-to-date to the conventions with less maintenance, however, as it's still marked as non-stable (current version is `1.0.0-rc9.9`), its inclusion is not included in this proposal.

Reviewer 3: ```suggestion
As its changes can be hard to follow, the .NET project includes a package named [`OpenTelemetry.SemanticConventions`](https://www.nuget.org/packages/OpenTelemetry.SemanticConventions/1.0.0-rc9.9) that maps the attributes defined in the conventions to a .NET project. Using this package will allow the Cassandra project to have its tracing attributes up-to-date to the conventions with less maintenance, however, as it's still marked as non-stable (current version is `1.0.0-rc9.9`), it is not included in this proposal.
```

- [SqlClient](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.SqlClient) (Community contribution)
- [MongoDB](https://github.com/jbogard/MongoDB.Driver.Core.Extensions.OpenTelemetry) (Community contribution)
- [Elasticsearch](https://github.com/elastic/elasticsearch-net/blob/main/src/Elastic.Clients.Elasticsearch/Client/ElasticsearchClient.cs#L183) (Native)

Cassandra also has client-side implementations in other languages in the form of contribution projects, as listed below:

- [Java](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/cassandra) (Community contribution)
- [NodeJS](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-cassandra) (Community contribution)
- [Python](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-cassandra) (Community contribution)

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- TBD

# Future possibilities
[future-possibilities]: #future-possibilities

## Traces

> ### Include missing Recommended attributes

Reviewer 3: ```suggestion
### Include missing recommended attributes
```
